### ISSUE TITLE
 Convert.ChangeType throws excpetion when converting Double.MaxValue

### ISSUE THREAD

**Issue by mdmoura** [dotnet/coreclr#5989]
The following code:

        var converted = Convert.ChangeType(Double.MaxValue.ToString(), typeof(Double));

Results in the exception:

       Value was either too large or too small for a Double.

I suppose this is a bug, correct?

**Comment by AlexGhiondea** [comment]
@mdmoura thanks for reporting this. It does seem strange that this throws. 

Would you be interested in looking at why that is and suggesting a solution?

Please note that this is the behavior we have on Desktop as well so any change should consider the compat impact as well.

**Comment by mazong1123** [comment]
In [number.cpp](https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/number.cpp#L564), if the most significant 3 hex number is greater than or equal to `0x7FF` , we'll hit the overflow exception:

```cpp
   else if (exp >= 0x7FF) {
        // overflow
        val = I64(0x7FF0000000000000);
    }
```

In fact, the Double.MaxValue is `0x7FFEFFFFFFFFFFFE` , which definitely enters into this trap. More over, try to convert any double number starts with `0x7FF` will hit this exception. There's [another code](https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/number.cpp#L2570) just check the first 3 hexes:

```cpp
if (e == 0x7FF) {
    FC_RETURN_BOOL(false);
}
```
I've no idea if it is a bug (probably yes) or by design. We have to change the approach for detecting the overflow of double if we wanna fixing this issue.

**Comment by AlexGhiondea** [comment]
@mazong1123 thanks for drilling into this! 

@jkotas @joshfree do you guys have any context on this? I assume we can come up with other ways to check for overflow if we believe that is the right solution.

**Comment by jkotas** [comment]
Floating point  ToString() values do not roundtrip. If you need value that roundtrips, you have to use the "R" formatting string. `Convert.ChangeType(Double.MaxValue.ToString("R"), typeof(Double));` works fine. It is how it has been in the framework since forever. I do not know why it was done that way - I agree that it is not a very intuitive design.

**Comment by jkotas** [comment]
cc @tarekgh since he has been working on similar issues recently

**Comment by mazong1123** [comment]
I think I've got more context about this issue when I'm working on #10651 . The [IEEE Double-precision floating-point format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) defined:

> The exponents 0x000 and 0x7ff have a special meaning:
0x000 is used to represent a signed zero (if F=0) and subnormals (if F≠0); and
0x7ff is used to represent ∞ (if F=0) and NaNs (if F≠0),

I think that's why the code said if the exponent is `0x7ff` we should throw overflow exception. I'll come back after #10651 resolved.

**Comment by mazong1123** [comment]
I think this issue is by design as `Double.MaxValue.ToString()` is not round-trippable as @jkotas said. Are we going to close this issue? @jkotas

**Comment by tarekgh** [comment]
CC @tannergooding

**Comment by tannergooding** [comment]
Right, we would have to take a breaking change to `double.ToString("G")` to fix this, which might be acceptable if we fixed `R` to always print the shortest round-trippable string. There are a few papers on this (and some other languages try to do this), but I haven't dug into it any deeper yet.

**Comment by mazong1123** [comment]
@tannergooding I believe following issues & PRs are related:

https://github.com/dotnet/coreclr/issues/13106
https://github.com/dotnet/corefx/pull/32268
https://github.com/dotnet/coreclr/pull/19905

I'm thinking if I can help on this issue. Due to my limited free time (only weekends), probably some minor or clean up jobs can assign to me so that @tannergooding may concentrate on the paper.
